apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,357 @@
+        /// </para>
+        /// </remarks>
+        /// <param name="tokenAccessMethod">The token access method to use. May be null.</param>
+        /// <returns>A builder with the specified token access method.</returns>
+        public TBuilder WithTokenAccessMethod(Func<string, CancellationToken, Task<string>> tokenAccessMethod)
+        {
+            var clone = Clone();
+            clone.TokenAccessMethod = tokenAccessMethod;
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified gRPC call invoker.
+        /// </summary>
+        /// <param name="callInvoker">The gRPC call invoker to use. May be null.</param>
+        /// <returns>A builder with the specified call invoker.</returns>
+        public TBuilder WithCallInvoker(CallInvoker callInvoker)
+        {
+            var clone = Clone();
+            clone.CallInvoker = callInvoker;
+            return clone;
+        }
+
+        /// <summary>
+        /// Validates that the builder is in a consistent state for building. For example, it's invalid to call
+        /// <see cref="Build"/> on an instance which has both JSON credentials and a credentials path specified.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The builder is in an invalid state.</exception>
+        protected virtual void Validate()
+        {
+            // If there's a call invoker, we shouldn't have any credentials-related information or an endpoint.
+            ValidateOptionExcludesOthers("CallInvoker cannot be specified with credentials settings or an endpoint", CallInvoker,
+                ChannelCredentials, CredentialsPath, JsonCredentials, Scopes, Endpoint, TokenAccessMethod);
+
+            ValidateAtMostOneNotNull("Only one source of credentials can be specified",
+                ChannelCredentials, CredentialsPath, JsonCredentials, TokenAccessMethod);
+
+            ValidateOptionExcludesOthers("Scopes are not relevant when a token access method or channel credentials are supplied", Scopes,
+                TokenAccessMethod, ChannelCredentials);            
+        }
+
+        /// <summary>
+        /// Validates that at most one of the given values is not null.
+        /// </summary>
+        /// <param name="message">The message if the condition is violated.</param>
+        /// <param name="values">The values to check for nullity.</param>
+        /// <exception cref="InvalidOperationException">More than one value is null.</exception>
+        protected void ValidateAtMostOneNotNull(string message, params object[] values)
+        {
+            int notNull = values.Count(v => v == null);
+            GaxPreconditions.CheckState(notNull < 2, message);
+        }
+
+        /// <summary>
+        /// Validates that if <paramref name="controlling"/> is not null, then no value in <paramref name="values"/> is null.
```

Comment:
nit: ... "then all values in value are null"

Reply:
Whoops, will fix.